beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -1725,13 +1733,98 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
+            HashSet::new();
+
+        let old_finalized_slot = self
+            .get_block(&old_finalized_block_hash.into())?
+            .ok_or_else(|| Error::MissingBeaconBlock(old_finalized_block_hash.into()))?
+            .message
+            .slot;
+
+        for (head_hash, _) in self.heads() {
+            let mut potentially_abandoned_blocks: HashSet<(
+                SignedBeaconBlockHash,
+                BeaconStateHash,
+                Slot,
+            )> = HashSet::new();
+
+            for (block_hash_, signed_beacon_block) in
+                ParentRootBlockIterator::new(&*self.store, head_hash)
+            {
+                let block_hash: SignedBeaconBlockHash = block_hash_.into();
+
+                if signed_beacon_block.message.slot > new_finalized_slot {
+                    // Tentatively schedule current block for deletion since it isn't know yet
+                    // whether the head is valid or not.
+                    potentially_abandoned_blocks.insert((
+                        block_hash,
+                        signed_beacon_block.message.state_root.into(),
+                        signed_beacon_block.message.slot,
+                    ));
+                } else if signed_beacon_block.message.slot == new_finalized_slot {
+                    if block_hash == new_finalized_block_hash {
+                        // The current head includes the newly finalized block, so we're either on
+                        // the canonical chain or on a legitimate candidate for the canonical
+                        // chain.  Do not remove any blocks.
+                        potentially_abandoned_blocks.clear();
+                    } else {
+                        // We've got a head which isn't based on the newly finalized block.  Delete
+                        // it and all its blocks.
+                        self.head_tracker.remove_head(head_hash);
+                        abandoned_blocks.extend(potentially_abandoned_blocks);
+                    }
+                    break;
+                } else {
+                    // We've got a block which doesn't include the newly finalized block. Moreover,
+                    // its slot number is lower than the newly finalized block's one. This means
+                    // this is a "neglected" fork created by some peers.  By now, it can't be
+                    // extended to include the newly finalized block since necessary slot numbers
+                    // are already occupied by the canonical chain.  Therefore, we can safely get
+                    // rid of this head and its blocks.
+
+                    abandoned_blocks.extend(
+                        ParentRootBlockIterator::new(&*self.store, head_hash)
+                            .take_while(|(_, signed_beacon_block)| {
+                                signed_beacon_block.message.slot > old_finalized_slot
```

Comment:
I think this case can be combined with the one above (we don't want to delete everything back to the `old_finalized_slot` because the abandoned chain could share blocks in common with the canonical chain)

Reply:
I've changed `ParentRootBlockIterator::new(&*self.store, head_hash)` into `ParentRootBlockIterator::new(&*self.store, block_hash.into())` to make it more apparent it's a separate case and can't really be combined with the above one as it handles slots lower than `new_finalized_slot`.

Reply:
Hmm, I think there's still an issue here, but I'm surprised the tests aren't catching it.

We can't take every block back to the old finalized slot because some of those blocks could be part of the canonical chain. Here's a diagram I made to illustrate this:

![Imgur](https://i.imgur.com/oiWfPfo.jpg)

We don't want to prune the blocks at slots `i + 4` or less, because although they're part of the abandoned chain, they're also part of the canonical chain. Pruning the blocks at slots `>i` will inadvertently delete them.

This is where I think we could use a `HashSet` of the block roots of all the blocks between the old finalized block and the new finalized block to work out when we've hit the finalized chain. Then the take would be:

```rust
.take_while(|(block_root, _)| !new_finalized_blocks.contains(block_root))
```

I think if you set the initial length of chain in the `prunes_abandoned_fork_between_two_finalized_checkpoints` test to be greater than `slots_per_epoch`, then the test will fail because of this issue.